Rubberduck.CodeAnalysis/Inspections/Abstract/ParseTreeInspectionBase.cs

```diff
@@ -1,15 +1,161 @@
+                this,
+                ResultDescription(context),
+                context,
+                DisabledQuickFixes(context));
+        }
+
+        protected virtual ICollection<string> DisabledQuickFixes(QualifiedContext<ParserRuleContext> context) => new List<string>();
+        public virtual CodeKind TargetKindOfCode => CodeKind.CodePaneCode;
+    }
+
+
+    public abstract class ParseTreeInspectionBase<T> : InspectionBase, IParseTreeInspection
+    {
+        protected ParseTreeInspectionBase(IDeclarationFinderProvider declarationFinderProvider)
+            : base(declarationFinderProvider)
+        {}
+
+        public abstract IInspectionListener Listener { get; }
+        protected abstract string ResultDescription(QualifiedContext<ParserRuleContext> context, T properties);
+        protected abstract (bool isResult, T properties) IsResultContextWithAdditionalProperties(QualifiedContext<ParserRuleContext> context);
+
+        protected override IEnumerable<IInspectionResult> DoGetInspectionResults()
+        {
+            return DoGetInspectionResults(Listener.Contexts());
+        }
+
+        protected override IEnumerable<IInspectionResult> DoGetInspectionResults(QualifiedModuleName module)
+        {
+            return DoGetInspectionResults(Listener.Contexts(module));
+        }
+
+        private IEnumerable<IInspectionResult> DoGetInspectionResults(IEnumerable<QualifiedContext<ParserRuleContext>> contexts)
+        {
+            var objectionableContexts = contexts
+                .Select(ContextsWithResultProperties)
+                .Where(result => result.HasValue)
+                .Select(result => result.Value);
+
+            return objectionableContexts
+                .Select(tpl => InspectionResult(tpl.context, tpl.properties))
+                .ToList();
+        }
+
+        private (QualifiedContext<ParserRuleContext> context, T properties)? ContextsWithResultProperties(QualifiedContext<ParserRuleContext> context)
+        {
+            var (isResult, properties) = IsResultContextWithAdditionalProperties(context);
+            return isResult
+                ? (context, properties)
+                : ((QualifiedContext<ParserRuleContext> context, T properties)?) null;
+        }
+
+        protected virtual IInspectionResult InspectionResult(QualifiedContext<ParserRuleContext> context, T properties)
+        {
+            return new QualifiedContextInspectionResult<T>(
+                this,
+                ResultDescription(context, properties),
+                context,
+                properties,
+                DisabledQuickFixes(context, properties));
+        }
+
+        protected virtual ICollection<string> DisabledQuickFixes(QualifiedContext<ParserRuleContext> context, T properties) => new List<string>();
         public virtual CodeKind TargetKindOfCode => CodeKind.CodePaneCode;
     }
+
+    public class InspectionListenerBase : VBAParserBaseListener, IInspectionListener
```

Comment:
Except for generic abstract classes inherited from a non-generic abstract class? 😉

Reply:
For parse tree inspections, we have always put the listeners into the same file.

I think, in many cases, we can actually make them private classes.

Reply:
I do agree with inlining the listeners in the implementation since most of the time they will be specific to only one inspection so it makes sense to make it a nested & private class. We do have to consider the discoverability aspect for the abstract classes, however. If I'm new to the code, I cannot see that there are both `ParseTreeInspectionBase` and `InspectionListenerBase` by viewing the filesystem alone. I'd have to use something like Find Implementation on the `IInspectionListener` to see that I need to create my listener based on the `InspectionListenerBase`. I am not sure off the cuff if the Find Implementation won't give cluttered results with all other concrete implementations.